ISSUE-165: plaintext and total sequence count capabilities to SBF Flavor Solr Documents

DiegoPino opened the pull request:
See #165 

# What is this

@giancarlobi @pcambra this works and is cleaner now. 

I created a "soft" dependency on a `Metadata Display Template` entity... in case the module is enabled it will be used. I will leave the code for a next pull there. Not used but will be used.

I also added the sequence total and now wonder if we should make both sequence and total integers instead of string? (for Solr).

In any case this can already be tested using the "matching" SBR https://github.com/giancarlobi/strawberry_runners/pull/33 pull.

 If used with already Tracked Strawberry Flavors, the total count will not be available. But I can even generate an Update hook and use Solr to "deduce that"  if it is needed for existing running instances. For new ones, using the "matching" pull in Strawberry Runners, all should be OK.

Please check and review. The fact that I merged a bug (not in production yet I hope) means this needs some urgency. Thanks

giancarlobi commented:
@DiegoPino I ingested a new ADO pdf and almost all works well. A couple of notes (I don't know if issues or my bad/missed something):
1. In Solr doc , I see that "ss_sequence_total" == "ss_sequence_id" and not is the total of pages
2. In Solr I have indexed plaintext (ok!) and 3 times miniocr as "tcocr_highlightm_X3b_und_fulltext_with_original_ocr", "tcocr_highlightm_X3b_und_ocr_text" and "tm_X3b_und_fulltext". I think we need only one, right? In addition the last one is using type "text_und" and that I think make no sense to store miniocr.

giancarlobi commented:
>     1. In Solr doc , I see that "ss_sequence_total" == "ss_sequence_id" and not is the total of pages

I checked that:
I have (I think) the right SBR settings 
![image](https://user-images.githubusercontent.com/4232178/112458470-1e9cce80-8d5d-11eb-8c9e-4a5f20f16c79.png)
I have flv:pdfinfo (1...13)
It seems no sequence_total present in input so `$sequence_total = isset($io->input->sequence_total) ? $io->input->sequence_total : $sequence_number;`

giancarlobi commented:
More, I added this `\Drupal::logger('ocr_runners')->notice($io->input->sequence_total);` at the beginning of `OcrPostProcessor.php` function `run` and variable returned empty/null.

giancarlobi commented:
@DiegoPino do we need to change something here https://github.com/esmero/strawberry_runners/blob/c8a1d39ba71b305393ea71f641537c8c96cb37ac/src/Plugin/StrawberryRunnersPostProcessor/OcrPostProcessor.php#L27 to include also sequence_total? sorry, I'm not so skilled about this

DiegoPino commented:
@giancarlobi. I updated https://github.com/giancarlobi/strawberry_runners/pull/33 and also did some cleanup here.
I see total sequence being indexed correctly for new HOCRs

![image](https://user-images.githubusercontent.com/6946023/112554442-b241bf80-8d9c-11eb-821a-77130c503780.png)

I also do not see any errors in my logs. I tried to keep the changes on the Strawberry runners module as discrete as possible ( I fear to break things there, the Queue worker logic is so complex but also so nice)

I also removed from my index all unnecessary fields for the Strawberry Flavor Data Source, means I kept only the ocr_text field, the parent title field (which leads to a question I will have for you later on... ), and added the total sequence. 


@pcambra I found some bugs in the Flavor Delete Subscriber (sorry,  I should have tested the code better) and proposed some solutions. I mentioned you in that particular commit so its easier to see what I did

Thanks. Please add your comments, code reviews, suggestions, etc. What a day

giancarlobi commented:
@DiegoPino A first check was ok: 
- I cleaned al Solr flavors and ADOs pdf to start from a clean point (from admin/config/search/search-api/index/default_solr_index, "Clear all indexed data" then "rebuild tracking..." then "Index now")
- ingested a new ADO pdf
- after all queues run, checked Solr doc flavor: all ok also the right sequece_total value
- no errors in dlog

Next steps will be: clean index of SBF flavor doc and check flavor delete.
Have a nice day

giancarlobi commented:
I cleaned index and made as your above. Ingested a new ADO pdf, all works well without error.
I have a couple of warning for each page/flavor probably due to some missed/bad config of my search api solr or some bed conf from previous test, just in case I report here:
- search_api: 
Message | The field Rendered HTML output (rendered_item)  on index Drupal Content to Solr 7 is missing view mode configuration  for some datasources or bundles. Please review (and re-save) the field  settings.
- search_api: 
Message | The field IIIF Image Tag for ADO  (rendered_item_iiif_img) on index Drupal Content to Solr 7 is missing  view mode configuration for some datasources or bundles. Please review  (and re-save) the field settings.
I have rendered_item and rendered_item_iiif_img defined into fields and configured with all default, just in case.

Now I'll check flavor delete.

giancarlobi commented:
@DiegoPino  Flavor delete, I have an issue, probably due to some my missed/bad configuration, I think.
I changed limit to 5.
I delete one of the previous ADO pdf and I see these errors in dlog:
- for three consecutive times : 
Message | Drupal\Core\Entity\Query\QueryException while  attempting to find indexed entities referencing changed Content with ID  "120" for index Drupal Content to Solr 7: Invalid specifier  'sbf_entity_reference_ismemberof' in  Drupal\Core\Entity\Query\Sql\Tables->addField() (line 306 of  /var/www/archipelago/web/core/lib/Drupal/Core/Entity/Query/Sql/Tables.php).

Then is logged as info 
Message | digital_object: deleted Test pdf 13.

Then (probably due to previous errors): 
Message | Could not load the following items on index Drupal Content to Solr 7: "entity:node/120:en".

I imagine that without solve this I cannot following with check,

giancarlobi commented:
In addition I checked delete also on the new one Archipelago and also in that case I receive the same search_api error for tree times.

DiegoPino commented:
Thanks Giancarlo. Let me wake up (7:45 second coffee) to look at that
deeper but my guess is it may be that search api patch you don’t have
because i do not get those errors in out deployments. There is a bug (or
feature hahaha) in search api that triggers errors if a entity reference
does not exist anymore. Can you share the full log? Also the code i fixed
only removes documents for deleted pdfs (means edit via webform, remove
file save) not for a full ado but will also add that one later today. See
you in a while!

El El vie, 26 de mar. de 2021 a la(s) 07:02, Giancarlo <
***@***.***> escribió:

> In addition I checked delete also on the new one Archipelago and also in
> that case I receive the same search_api error for tree times.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/esmero/strawberryfield/pull/168#issuecomment-808123209>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABU7ZZ455CI2YE33EC6UQ6DTFRSV7ANCNFSM4ZYDQEGA>
> .
>
-- 
Diego Pino Navarro
Digital Repositories Developer
Metropolitan New York Library Council (METRO)

giancarlobi commented:
@DiegoPino I managed pdf by webform, in both case.
The full log is :

Type | search_api
-- | --
Friday, March 26, 2021 - 09:53
ArchiSiteAdmin
https://archipelago.byterfly.eu/node/120/delete
https://archipelago.byterfly.eu/node/120/delete
Drupal\Core\Entity\Query\QueryException while  attempting to find indexed entities referencing changed Content with ID  "120" for index Drupal Content to Solr 7: Invalid specifier  'sbf_entity_reference_ismemberof' in  Drupal\Core\Entity\Query\Sql\Tables->addField() (line 306 of  /var/www/archipelago/web/core/lib/Drupal/Core/Entity/Query/Sql/Tables.php).
Error
150.145.55.50
Go to changed Content with ID "120"

Is this ok?
And for more info:
```
composer show drupal/search_api
[sudo] password for giancarlo:
name     : drupal/search_api
descrip. : Provides a generic framework for modules offering search capabilities.
keywords :
versions : * 1.19.0
type     : drupal-module
license  : GNU General Public License v2.0 or later (GPL-2.0-or-later) (OSI approved) https://spdx.org/licenses/GPL-2.0-or-later.html#licenseText
homepage : https://www.drupal.org/project/search_api
source   : [git] https://git.drupalcode.org/project/search_api.git 8.x-1.19
dist     : [zip] https://ftp.drupal.org/files/projects/search_api-8.x-1.19.zip 8.x-1.19
path     : /var/www/archipelago/web/modules/contrib/search_api
names    : drupal/search_api
```
```
composer show drupal/search_api_solr
name     : drupal/search_api_solr
descrip. : Offers an implementation of the Search API that uses an Apache Solr server for indexing content.
keywords :
versions : * 4.1.11
type     : drupal-module
license  : GNU General Public License v2.0 or later (GPL-2.0-or-later) (OSI approved) https://spdx.org/licenses/GPL-2.0-or-later.html#licenseText
homepage : https://www.drupal.org/project/search_api_solr
source   : [git] https://git.drupalcode.org/project/search_api_solr.git 4.1.11
dist     : [zip] https://ftp.drupal.org/files/projects/search_api_solr-4.1.11.zip 4.1.11
path     : /var/www/archipelago/web/modules/contrib/search_api_solr
names    : drupal/search_api_solr
```
Lunch time here, see you later, friend.

giancarlobi commented:
> search api patch you don’t have because i do not get those errors in out deployments.

@DiegoPino Do you mean this? 
```
"drupal/search_api": {
                "Fix Content Entity Tracker in 1.18": "patches/1.18-3178941-metrocustom-referenced_entities_reindexing_fatal_error-D8.patch"
            },
```

DiegoPino commented:
Thanks a lot. That helps. Will debug asap

El El vie, 26 de mar. de 2021 a la(s) 08:09, Giancarlo <
***@***.***> escribió:

> @DiegoPino <https://github.com/DiegoPino> I managed pdf by webform, in
> both case.
> The full log is :
> Type search_api
> Friday, March 26, 2021 - 09:53
> ArchiSiteAdmin
> https://archipelago.byterfly.eu/node/120/delete
> https://archipelago.byterfly.eu/node/120/delete
> Drupal\Core\Entity\Query\QueryException while attempting to find indexed
> entities referencing changed Content with ID "120" for index Drupal Content
> to Solr 7: Invalid specifier 'sbf_entity_reference_ismemberof' in
> Drupal\Core\Entity\Query\Sql\Tables->addField() (line 306 of
> /var/www/archipelago/web/core/lib/Drupal/Core/Entity/Query/Sql/Tables.php).
> Error
> 150.145.55.50
> Go to changed Content with ID "120"
>
> Is this ok?
> And for more info:
>
> composer show drupal/search_api
> [sudo] password for giancarlo:
> name     : drupal/search_api
> descrip. : Provides a generic framework for modules offering search capabilities.
> keywords :
> versions : * 1.19.0
> type     : drupal-module
> license  : GNU General Public License v2.0 or later (GPL-2.0-or-later) (OSI approved) https://spdx.org/licenses/GPL-2.0-or-later.html#licenseText
> homepage : https://www.drupal.org/project/search_api
> source   : [git] https://git.drupalcode.org/project/search_api.git 8.x-1.19
> dist     : [zip] https://ftp.drupal.org/files/projects/search_api-8.x-1.19.zip 8.x-1.19
> path     : /var/www/archipelago/web/modules/contrib/search_api
> names    : drupal/search_api
>
> composer show drupal/search_api_solr
> name     : drupal/search_api_solr
> descrip. : Offers an implementation of the Search API that uses an Apache Solr server for indexing content.
> keywords :
> versions : * 4.1.11
> type     : drupal-module
> license  : GNU General Public License v2.0 or later (GPL-2.0-or-later) (OSI approved) https://spdx.org/licenses/GPL-2.0-or-later.html#licenseText
> homepage : https://www.drupal.org/project/search_api_solr
> source   : [git] https://git.drupalcode.org/project/search_api_solr.git 4.1.11
> dist     : [zip] https://ftp.drupal.org/files/projects/search_api_solr-4.1.11.zip 4.1.11
> path     : /var/www/archipelago/web/modules/contrib/search_api_solr
> names    : drupal/search_api_solr
>
> Lunch time here, see you later, friend.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/esmero/strawberryfield/pull/168#issuecomment-808166440>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABU7ZZ44DKDT7WHGS7FDLL3TFR2OBANCNFSM4ZYDQEGA>
> .
>
-- 
Diego Pino Navarro
Digital Repositories Developer
Metropolitan New York Library Council (METRO)

DiegoPino commented:
@giancarlobi there is an update there do deal with #169 . Now also the Source Key Values will be removed when untracked. I feel its in general a bit drastic since you could want to revert a revision shortly after making a "silly" change like deleting a file, but then it is also safer. In the future we could set the "removal" into a queue, which would only run via Hydroponics? Giving the user the chance to do some CTRL-Z. But for now I'm happy with it. Please read the last Commit comments about expected behavior in the Batch Indexer.  I also updated https://github.com/esmero/strawberry_runners/pull/33 to use a Constant defined there for the Key Value Collection. Let me know what you think.

giancarlobi commented:
@DiegoPino I fully agree with you about your choice to remove Source Key Values when untracked, I think it's the safer mode for now, in the future we can change this behavior but at the moment this makes all working well without errors and duplicated flavors.

DiegoPino commented:
@giancarlobi merged. Looked at the code a few times and tested with real data. I see no issues right now. We will keep iterating on making the data source better. Thanks for testing
